website/content/docs/nia/terraform-modules.mdx

```diff
@@ -0,0 +1,123 @@
+2. **`services` input variable** - Consul-Terraform-Sync requires all modules to have the following input variable declared within the root module. The declaration of the `services` variable can be included at the top of the suggested `variables.tf` file where other input variables are commonly declared. This variable functions as the response object from the Consul catalog API and surfaces network information to be consumed by the module. It is structured as a map of objects.
+
+```hcl
+variable "services" {
+  description = "Consul services monitored by Consul-Terraform-Sync"
+  type = map(
+    object({
+      id        = string
+      name      = string
+      kind      = string
+      address   = string
+      port      = number
+      meta      = map(string)
+      tags      = list(string)
+      namespace = string
+      status    = string
+
+      node                  = string
+      node_id               = string
+      node_address          = string
+      node_datacenter       = string
+      node_tagged_addresses = map(string)
+      node_meta             = map(string)
+
+      cts_user_defined_meta = map(string)
+    })
+  )
+}
+```
+
+Keys of the `services` map are unique identifiers of the service across Consul agents and data centers. Keys follow the format `service-id.node.datacenter` (or `service-id.node.namespace.datacenter` for Consul Enterprise). A complete list of attributes available for the `services` variable is included in the [documentation for Consul-Terraform-Sync tasks](/docs/nia/tasks#task-execution).
+
+## How to Create a Compatible Terraform Module
+
+You can read more on how to [create a module](https://www.terraform.io/docs/modules/index.html) or work through a [tutorial to build a module](https://learn.hashicorp.com/tutorials/terraform/module-create). Consul-Terraform-Sync is designed to integrate with any module that satisfies the specifications in the following section.
+
+The repository [hashicorp/consul-terraform-sync-template-module](https://github.com/hashicorp/consul-terraform-sync-template-module) can be cloned and used as a starting point for structuring a compatible Terraform module.
+
+First, create a directory to organize Terraform configuration files that make up the module. You can start of with creating two files `main.tf` and `variables.tf` and expand from there based on your module and network infrastructure automation needs. The `main.tf` is the entry point of the module and this is where you can begin authoring your module. It can contain multiple Terraform resources related to an automation task that uses Consul service discovery information, particularly the required `services` input variable.
+
+The code example below shows a resource using the `services` variable. When this example is used in automation with Consul-Terraform-Sync, the content of the local file would dynamically update as Consul service discovery information changes.
```

Comment:
Since the services variable declaration has language like 'copy this to variables.tf', suggestion that you don't have to take to have parity here with similar 'copy this to the main.tf'

Reply:
it's a bit different since there's not a simple approach to copy and paste for `main.tf`. so instead i reorganized it to better clarify the differences, hope it helps! lmk what you think

Reply:
https://consul-git-cts-tf-modules-hashicorp.vercel.app/docs/nia/terraform-modules#how-to-create-a-compatible-terraform-module

Reply:
Yes! Having the separate services variable section helps showing how these two files are handled and talked about differently. Thanks!